IbFlexReader/IbFlexReader/Reader.cs

```diff
@@ -2,92 +2,215 @@
-
-                    if (response.Status != "Success")
-                    {
-                        return new FlexResult
-                        {
-                            ErrorCode = response.ErrorCode,
-                            ErrorMessage = response.ErrorMessage
-                        };
-                    }
-
-                    referenceCode = response.ReferenceCode.Value.ToString();
-                }
-
-
-                uri = new Uri($"https://gdcdyn.interactivebrokers.com/Universal/servlet/FlexStatementService.GetStatement?q={referenceCode}&t={token}&v=3");
-
-                using (var getStatementResult = await client.PostAsync(uri, null))
-                using (var stream = await getStatementResult.Content.ReadAsStreamAsync())
+        public async Task<FlexResult> GetByApi(string token, string queryId, string dumpFile = null, int? retry = null)
+		{
+			using (var client = new HttpClient())
+			{
+				FlexResult requestResult = await SendRequest(client, token, queryId, dumpFile);
+
+				if (!requestResult.IsSuccess)
+				{
+					return requestResult;
+				}
+
+				string referenceCode = requestResult.ErrorMessage;
+
+				int retryCounter = 0;
+				if (retry.HasValue && retry.Value > 0) retryCounter = retry.Value;
+
+				FlexResult result = null;
+
+				bool tryAgain = true;
+				while (tryAgain)
+				{
+					result = await GetStatement(client, token, referenceCode, dumpFile);
+
+					if (!result.IsSuccess && result.ErrorCode.HasValue && result.ErrorCode.Value == 1019)
+					{
+                        // warning is: 'Statement generation in progress. Please try again shortly. '
+
+                        System.Diagnostics.Debug.Print("IbFlexReader.Reader.GetByApi: Retry");
+
+						// retry getting data?
+						if (retryCounter > 0)
+						{
+							// yes
+							retryCounter--;
+                            // wait a little bit 
+							System.Threading.Thread.Sleep(2000);
+						}
+						else
+						{
+							// no
+							tryAgain = false;
+						}
+					}
+					else
+					{
+						// data received
+						tryAgain = false;
+					}
+				}
+
+				return result;
+
+			}
+		}
+
+		private async Task<FlexResult> GetStatement(HttpClient client, string token, string referenceCode, string dumpFile)
+		{
+			var uri = new Uri($"https://gdcdyn.interactivebrokers.com/Universal/servlet/FlexStatementService.GetStatement?q={referenceCode}&t={token}&v=3");
+
+			using (var getStatementResult = await client.PostAsync(uri, null))
+			using (var stream = await getStatementResult.Content.ReadAsStreamAsync())
+			{
+				if (dumpFile != null) { Utils.DumpFileWriter.DumpGetStatement(stream, dumpFile); }
+
+				// in case of an error or warning, a FlexStatementResponse object will be returned instead of the data
+				if (TryGetFlexStatementResponse(stream, out FlexStatementResponse flexStatementResponse))
```

Comment:
Isn't this IF returning for succesfull response fetches? I think you're missing an exclamation mark here^^

Reply:
I think this is right (and it works proper in my code)
You first send _FlexStatementService.SendRequest_ and IB returns a **FlexStatementResponse** (see line 159ff in  method **SendRequest**.
If no error or warning occurs you get the ReferenceCode which you use in your next call to IB which is _FlexStatementService.GetStatement_.

Now there are two possibilities:

1. the call is successfull and IB sends the data. Then the result is a **FlexQueryResponse**
2. the call results in error or warning. Then the result is a **FlexStatementResponse**

Line 113 checks if the result is a **FlexStatementResponse**, if so we have an error or warning and exit the method.
If the result is not a **FlexStatementResponse** then it should be a **FlexQueryResponse** containing the data. Then the data is deserialized in line 123.

Reply:
Ups. you're right. ^^